### DISCUSSION TITLE
TryCatch circles in C#

### DISCUSSION THREAD

**Discussion by BartBurton** [dotnet/csharplang#5971] [Language Ideas] [answered]
Hey, how you an idea - add **retry** operator for **try{...} catch{...}** blocks. That may be a key word like **retry;** in **catch{...}** block that execute **try{...}** block operators again. And we'll can write something like - **retry;**_(infinity)_, **retry once;**, **retry twice;**, **retry 3;** and so on.

**For example:**
_Was:_

```
TcpClient client = new(); 
int tries = 0, maxTries = 3; 
while (true) 
{ 
    try 
    { 
        client.Connect(ip, port); 
        break;
    } 
    catch (Exception e) 
    { 
        if (++tries == maxTries) throw; 
    } 
}
```

_Could be:_

```
TcpClient client = new(); 
try 
{ 
    client.Connect(ip, port); 
} 
catch (Exception e) 
{ 
    retry 3;
    throw; 
}
```

**Comment by LoneWald** [comment]
I realy need this, tnx.

**Comment by YairHalberstadt** [comment]
This is trivial to write a function for:

```csharp
void Retry(Action a, int? retries = null)
```

Also retrying is generally much more complex than just do this thing 3 times. You need to consider checking the type of error, back off, logging, etc. It shouldn't be something that is so trivial that it requires a language feature.

**Comment by BartBurton** [comment]
I still think it will be not bad
_See that example:_
```
TcpClient client = new(); 
try 
{ 
    client.Connect(ip, port); 
} 
catch (SocketException e) 
{ 
    Logger.WriteLogs(e.Message);
    retry 3;
    throw; 
}
catch
{
    throw; 
}
```

**Reply by YairHalberstadt** [reply]
Are you going to just retry 3 times? No back off? No configuration?

And why is this so common and important that you need a language feature?

The area of retrying is complex enough that there's entire libraries (e.g. Polly) dedicated to doing this correctly.

Remember every language feature starts in a hole of -100 points. It needs an incredibly strong reason to say that this can't be implemented in a library, and absolutely requires language support for it to be done.

**Comment by BartBurton** [comment]
This is just syntax sugar. We already have key words for linq)))

**Comment by BartBurton** [comment]
Try{} will be executed while any statement throws an exception, if it occurs three times, then the statements after **retry** in the Catch{} block are executed, and at each iteration, the statements before **retry** are executed, if on any then the iteration had no exception, then the function runs further. Cool

**Comment by HaloFour** [comment, marked as answer]
This is trivially accomplished without a language feature for it, especially one that will almost always be the wrong thing to do.  Given the amount of policy implies with retry semantics that would be better served through a library.
